File: backends/go/internal/usecase/payment.go
package usecase

import (
	"context"

	"github.com/ticketing-labs/backend-go/internal/domain"
)

// PaymentGateway is the external provider (the fake one in this lab). Charge asks it
// to settle an order; the gateway later calls our webhook with the result.
type PaymentGateway interface {
	Charge(ctx context.Context, orderID string) (providerRef string, err error)
}

type PaymentService struct {
	orders       OrderRepository
	reservations ReservationRepository
	payments     PaymentRepository
	gateway      PaymentGateway
	ids          IDGenerator
}

func NewPaymentService(o OrderRepository, r ReservationRepository, p PaymentRepository, gw PaymentGateway, ids IDGenerator) *PaymentService {
	return &PaymentService{orders: o, reservations: r, payments: p, gateway: gw, ids: ids}
}

// ProcessPaymentRequest is called by the broker worker for each payment.requested
// message. It records a pending payment and asks the gateway to charge. The result
// arrives asynchronously via the webhook.
func (s *PaymentService) ProcessPaymentRequest(ctx context.Context, orderID string) error {
	order, err := s.orders.FindByID(ctx, orderID)
	if err != nil || order == nil {
		return domain.ErrNotFound
	}
	if order.Status != domain.OrderPending {
		return nil // already settled; the message is a duplicate, ignore it
	}
	providerRef, err := s.gateway.Charge(ctx, orderID)
	if err != nil {
		return err // the worker will retry with backoff
	}
	return s.payments.Upsert(ctx, &domain.Payment{
		ID:          s.ids.NewID(),
		OrderID:     orderID,
		ProviderRef: providerRef,
		Status:      domain.PaymentPending,
		Attempts:    1,
	})
}

// HandleWebhook settles an order from a verified provider callback. Signature
// verification happens at the transport edge before this is called. Idempotent by
// provider_ref: a replayed webhook does not double-confirm anything.
func (s *PaymentService) HandleWebhook(ctx context.Context, providerRef, orderID string, succeeded bool) error {
	order, err := s.orders.FindByID(ctx, orderID)
	if err != nil || order == nil {
		return domain.ErrNotFound
	}

	status := domain.PaymentFailed
	if succeeded {
		status = domain.PaymentSucceeded
	}
	if err := s.payments.Upsert(ctx, &domain.Payment{
		ID:          s.ids.NewID(),
		OrderID:     orderID,
		ProviderRef: providerRef,
		Status:      status,
	}); err != nil {
		return domain.ErrInternal
	}

	if order.Status != domain.OrderPending {
		return nil // already settled; nothing more to do
	}

	if succeeded {
		if err := s.orders.UpdateStatus(ctx, orderID, domain.OrderPaid); err != nil {
			return domain.ErrInternal
		}
		// The one place the two state machines touch: a paid order confirms its hold.
		return s.reservations.UpdateStatus(ctx, order.ReservationID, domain.ReservationConfirmed)
	}
	return s.orders.UpdateStatus(ctx, orderID, domain.OrderFailed)
}

File: backends/fastapi/app/usecase/payment.py
from __future__ import annotations

from app.domain import errors
from app.domain.models import OrderStatus, Payment, PaymentStatus, ReservationStatus
from app.usecase.ports import (
    IDGenerator,
    OrderRepository,
    PaymentGateway,
    PaymentRepository,
    ReservationRepository,
)


class PaymentService:
    def __init__(
        self,
        orders: OrderRepository,
        reservations: ReservationRepository,
        payments: PaymentRepository,
        gateway: PaymentGateway,
        ids: IDGenerator,
    ):
        self._orders = orders
        self._reservations = reservations
        self._payments = payments
        self._gateway = gateway
        self._ids = ids

    async def process_payment_request(self, order_id: str) -> None:
        """Called by the broker worker per payment.requested message. Records a pending
        payment and asks the gateway to charge; the result arrives via the webhook.
        """
        order = await self._orders.find_by_id(order_id)
        if order is None:
            raise errors.NOT_FOUND
        if order.status != OrderStatus.PENDING:
            return  # already settled; duplicate message, ignore
        provider_ref = await self._gateway.charge(order_id)  # may raise; worker retries
        await self._payments.upsert(
            Payment(
                id=self._ids.new_id(),
                order_id=order_id,
                provider_ref=provider_ref,
                status=PaymentStatus.PENDING,
                attempts=1,
            )
        )

    async def handle_webhook(self, provider_ref: str, order_id: str, succeeded: bool) -> None:
        """Settle an order from a verified provider callback. Signature verification
        happens at the transport edge. Idempotent by provider_ref.
        """
        order = await self._orders.find_by_id(order_id)
        if order is None:
            raise errors.NOT_FOUND

        status = PaymentStatus.SUCCEEDED if succeeded else PaymentStatus.FAILED
        await self._payments.upsert(
            Payment(id=self._ids.new_id(), order_id=order_id, provider_ref=provider_ref, status=status)
        )

        if order.status != OrderStatus.PENDING:
            return  # already settled

        if succeeded:
            await self._orders.update_status(order_id, OrderStatus.PAID.value)
            # The one place the two state machines touch: a paid order confirms its hold.
            await self._reservations.update_status(order.reservation_id, ReservationStatus.CONFIRMED.value)
        else:
            await self._orders.update_status(order_id, OrderStatus.FAILED.value)
